From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 15 Aug 2025 16:16:08 +0900
Subject: [PATCH] Paper PR: Optimise temptation lookups changes

Some changes to the original temptation lookups optimization patch
- Replaced some storages from ArrayList to Array to be faster
- Cached mob location access in iteration
- Added tempt goal lookup optimizations for:
    - Purpur's cow mushrooms feed

diff --git a/io/papermc/paper/entity/temptation/GlobalTemptationLookup.java b/io/papermc/paper/entity/temptation/GlobalTemptationLookup.java
index 1f206998a61af887e312953319e69fc79e10e7e1..b38f1412b6c524c9f19f7e1cd7b532303459e5a3 100644
--- a/io/papermc/paper/entity/temptation/GlobalTemptationLookup.java
+++ b/io/papermc/paper/entity/temptation/GlobalTemptationLookup.java
@@ -18,6 +18,7 @@ public class GlobalTemptationLookup {
     public static final TemptationPredicate BEE_FOOD = register(stack -> stack.is(ItemTags.BEE_FOOD));
     public static final TemptationPredicate CHICKEN_FOOD = register(stack -> stack.is(ItemTags.CHICKEN_FOOD));
     public static final TemptationPredicate COW_FOOD = register(stack -> stack.is(ItemTags.COW_FOOD));
+    public static final TemptationPredicate COW_FOOD_MUSHROOM = register(stack -> stack.is(Items.RED_MUSHROOM) || stack.is(Items.BROWN_MUSHROOM) || stack.is(ItemTags.COW_FOOD)); // Leaf - Paper PR: Optimise temptation lookups changes
     public static final TemptationPredicate PANDA_FOOD = register(stack -> stack.is(ItemTags.PANDA_FOOD));
     public static final TemptationPredicate PIG_CARROT_ON_A_STICK = register(stack -> stack.is(Items.CARROT_ON_A_STICK));
     public static final TemptationPredicate PIG = register(stack -> stack.is(ItemTags.PIG_FOOD));
@@ -57,21 +58,37 @@ public class GlobalTemptationLookup {
         return val;
     }
 
-    private final List<BitSet> precalculatedTemptItems = new ArrayList<>();
+    // Leaf start - Paper PR: Optimise temptation lookups changes
+    private final BitSet[] precalculatedTemptItems;
     private final BitSet calculatedThisTick = new BitSet();
+    private static final net.minecraft.server.level.ServerPlayer[] EMPTY_PLAYERS = {};
+    private net.minecraft.server.level.ServerPlayer[] players = EMPTY_PLAYERS;
 
     {
-        for (int i = 0; i < registeredPredicateCounter; i++) {
-            this.precalculatedTemptItems.add(new BitSet());
+        precalculatedTemptItems = new BitSet[registeredPredicateCounter];
+        for (int i = 0; i < precalculatedTemptItems.length; i++) {
+            this.precalculatedTemptItems[i] = new BitSet();
         }
     }
 
-    public void reset() {
+    public void tick(final net.minecraft.server.level.ServerLevel world) {
         for (int i = 0; i < registeredPredicateCounter; i++) {
-            this.precalculatedTemptItems.get(i).clear();
+            this.precalculatedTemptItems[i].clear();
+            this.calculatedThisTick.clear();
+        }
+
+        int j = 0;
+        final net.minecraft.server.level.ServerPlayer[] array = world.players().toArray(EMPTY_PLAYERS);
+        for (int i = 0; i < array.length; i++) {
+            final net.minecraft.server.level.ServerPlayer p = array[i];
+            if (!p.isSpectator() && p.isAlive()) {
+                array[j] = p;
+                j++;
+            }
         }
-        this.calculatedThisTick.clear();
+        this.players = it.unimi.dsi.fastutil.objects.ObjectArrays.setLength(array, j);
     }
+    // Leaf end - Paper PR: Optimise temptation lookups changes
 
     public boolean isCalculated(final int index) {
         return this.calculatedThisTick.get(index);
@@ -82,6 +99,12 @@ public class GlobalTemptationLookup {
     }
 
     public BitSet getBitSet(final int index) {
-        return this.precalculatedTemptItems.get(index);
+        // Leaf start - Paper PR: Optimise temptation lookups changes
+        return this.precalculatedTemptItems[index];
+    }
+
+    public net.minecraft.server.level.ServerPlayer[] players() {
+        return this.players;
     }
+    // Leaf end - Paper PR: Optimise temptation lookups changes
 }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 696354ce6e9c1df6d8b911773aa7033ce2bc8723..b8b0415ca3ba428de0836498a4b32f443e747d8f 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1000,7 +1000,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
             boolean didDespawn = tickRateManager.runsNormally() && despawnMap.tick(this, this.entityTickList); // Leaf - optimize despawn
-            this.globalTemptationLookup.reset(); // Paper - optimise temptation lookups - reset global cache prior to next entity tick
+            this.globalTemptationLookup.tick(this); // Paper - optimise temptation lookups - reset global cache prior to next entity tick // Leaf - Paper PR: Optimise temptation lookups changes
             this.entityTickList
                 .forEach(
                     entity -> {
diff --git a/net/minecraft/world/entity/ai/goal/TemptGoal.java b/net/minecraft/world/entity/ai/goal/TemptGoal.java
index 935376900a41c00f95b3b42e40242ebed1409eb8..ce27e3d21ffd4d175c7349943499f52683abc870 100644
--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -63,21 +63,27 @@ public class TemptGoal extends Goal {
                 final net.minecraft.server.level.ServerLevel level = getServerLevel(this.mob);
                 final io.papermc.paper.entity.temptation.GlobalTemptationLookup lookup = level.getTemptGoalLookup();
                 final java.util.BitSet lookupBitSet = lookup.getBitSet(this.globalTemptationLookupIndex);
-                final java.util.List<net.minecraft.server.level.ServerPlayer> players = level.players();
+                final net.minecraft.server.level.ServerPlayer[] players = lookup.players(); // Leaf - Paper PR: Optimise temptation lookups changes
                 // Check if the lookup needs to be computed this tick. Do so for all players if needed.
                 if (!lookup.isCalculated(this.globalTemptationLookupIndex)) {
-                    for (int i = 0; i < players.size(); i++) {
-                        lookupBitSet.set(i, shouldFollow(players.get(i)));
+                    for (int i = 0; i < players.length; i++) { // Leaf - Paper PR: Optimise temptation lookups changes
+                        lookupBitSet.set(i, shouldFollow(players[i])); // Leaf - Paper PR: Optimise temptation lookups changes
                     }
                     lookup.setCalculated(this.globalTemptationLookupIndex);
                 }
                 double d = -1.0;
                 net.minecraft.server.level.ServerPlayer nearestPlayer = null;
                 // Only iterate over players that passed #shouldFollow either in the prior computation or another goals canUse check.
+                // Leaf start - Paper PR: Optimise temptation lookups changes
+                final Mob mob = this.mob;
+                final double mobX = mob.getX();
+                final double mobY = mob.getY();
+                final double mobZ = mob.getZ();
                 for (int i = lookupBitSet.nextSetBit(0); i >= 0; i = lookupBitSet.nextSetBit(i + 1)) {
-                    final net.minecraft.server.level.ServerPlayer player = players.get(i);
-                    if (rangeTargetingConditions.test(level, this.mob, player)) {
-                        final double d1 = player.distanceToSqr(this.mob.getX(), this.mob.getY(), this.mob.getZ());
+                    final net.minecraft.server.level.ServerPlayer player = players[i];
+                    if (rangeTargetingConditions.test(level, mob, player)) {
+                        final double d1 = player.distanceToSqr(mobX, mobY, mobZ);
+                        // Leaf end - Paper PR: Optimise temptation lookups changes
                         if (d == -1.0 || d1 < d) {
                             d = d1;
                             nearestPlayer = player;
diff --git a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
index cb5bb5ae6e4de25e0de0bbd24e8a35bc2333658c..1bf9d8bc7addb152c4ca025cf86f8229f3fec5ed 100644
--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -46,11 +46,11 @@ public class TemptingSensor extends Sensor<PathfinderMob> {
         if (this.globalTemptationLookupIndex != -1) {
             final io.papermc.paper.entity.temptation.GlobalTemptationLookup lookup = level.getTemptGoalLookup();
             final java.util.BitSet lookupBitSet = lookup.getBitSet(this.globalTemptationLookupIndex);
-            final java.util.List<net.minecraft.server.level.ServerPlayer> players = level.players();
+            final net.minecraft.server.level.ServerPlayer[] players = lookup.players(); // Leaf - Paper PR: Optimise temptation lookups changes
             // Check if the lookup needs to be computed this tick. Do so for all players if needed.
             if (!lookup.isCalculated(this.globalTemptationLookupIndex)) {
-                for (int i = 0; i < players.size(); i++) {
-                    final net.minecraft.server.level.ServerPlayer serverPlayer = players.get(i);
+                for (int i = 0; i < players.length; i++) { // Leaf - Paper PR: Optimise temptation lookups changes
+                    final net.minecraft.server.level.ServerPlayer serverPlayer = players[i]; // Leaf - Paper PR: Optimise temptation lookups changes
                     lookupBitSet.set(i, net.minecraft.world.entity.EntitySelector.NO_SPECTATORS.test(serverPlayer) && this.playerHoldingTemptation(entity, serverPlayer)); // check on update
                 }
                 lookup.setCalculated(this.globalTemptationLookupIndex);
@@ -58,10 +58,15 @@ public class TemptingSensor extends Sensor<PathfinderMob> {
             double d = -1.0;
             net.minecraft.server.level.ServerPlayer nearestPlayer = null;
             // Only iterate over players that passed #shouldFollow either in the prior computation or another goals canUse check.
+            // Leaf start - Paper PR: Optimise temptation lookups changes
+            final double entityX = entity.getX();
+            final double entityY = entity.getY();
+            final double entityZ = entity.getZ();
             for (int i = lookupBitSet.nextSetBit(0); i >= 0; i = lookupBitSet.nextSetBit(i + 1)) {
-                final net.minecraft.server.level.ServerPlayer player = players.get(i);
+                final net.minecraft.server.level.ServerPlayer player = players[i];
                 if (targetingConditions.test(level, entity, player) && !entity.hasPassenger(player)) { // check on update - consider non passengers
-                    final double d1 = player.distanceToSqr(entity.getX(), entity.getY(), entity.getZ());
+                    final double d1 = player.distanceToSqr(entityX, entityY, entityZ);
+                    // Leaf end - Paper PR: Optimise temptation lookups changes
                     if (d == -1.0 || d1 < d) {
                         d = d1;
                         nearestPlayer = player;
diff --git a/net/minecraft/world/entity/animal/cow/AbstractCow.java b/net/minecraft/world/entity/animal/cow/AbstractCow.java
index d77fb6f0dd541e37367938b3b5ada2aeb77dfcd9..2c807ec53cd1e9df229e51711f9df6c233f86aaa 100644
--- a/net/minecraft/world/entity/animal/cow/AbstractCow.java
+++ b/net/minecraft/world/entity/animal/cow/AbstractCow.java
@@ -41,7 +41,7 @@ public abstract class AbstractCow extends Animal {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(1, new PanicGoal(this, 2.0));
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, io.papermc.paper.entity.temptation.GlobalTemptationLookup.COW_FOOD, false)); // Purpur - Cows eat mushrooms // Paper - optimise temptation lookups
+        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, level().purpurConfig.cowFeedMushrooms > 0 ? io.papermc.paper.entity.temptation.GlobalTemptationLookup.COW_FOOD_MUSHROOM : io.papermc.paper.entity.temptation.GlobalTemptationLookup.COW_FOOD, false)); // Purpur - Cows eat mushrooms // Paper - optimise temptation lookups // Leaf - Paper PR: Optimise temptation lookups changes
         this.goalSelector.addGoal(4, new FollowParentGoal(this, 1.25));
         this.goalSelector.addGoal(5, new WaterAvoidingRandomStrollGoal(this, 1.0));
         this.goalSelector.addGoal(6, new LookAtPlayerGoal(this, Player.class, 6.0F));
